module/os/linux/zfs/zpl_xattr.c

```diff
@@ -706,19 +704,33 @@ static int
 __zpl_xattr_user_get(struct inode *ip, const char *name,
     void *value, size_t size)
 {
-	char *xattr_name;
+	boolean_t compat = !!(ITOZSB(ip)->z_flags & ZSB_XATTR_COMPAT);
+	boolean_t fallback = !!(ITOZSB(ip)->z_flags & ZSB_XATTR_FALLBACK);
 	int error;
 	/* xattr_resolve_name will do this for us if this is defined */
 #ifndef HAVE_XATTR_HANDLER_NAME
 	if (strcmp(name, "") == 0)
 		return (-EINVAL);
 #endif
+	if (ZFS_XA_NS_PREFIX_FORBIDDEN(name))
+		return (-EINVAL);
 	if (!(ITOZSB(ip)->z_flags & ZSB_XATTR))
 		return (-EOPNOTSUPP);
 
-	xattr_name = kmem_asprintf("%s%s", XATTR_USER_PREFIX, name);
-	error = zpl_xattr_get(ip, xattr_name, value, size);
-	kmem_strfree(xattr_name);
+	/*
+	 * Try to look up the name without the namespace prefix first for
+	 * compatibility with xattrs from other platforms.  If that fails,
+	 * try again with the namespace prefix if fallback is allowed.
+	 */
+	error = -ENODATA;
+	if (compat || fallback)
+		error = zpl_xattr_get(ip, name, value, size);
+	if (!compat || (fallback && error == -ENODATA)) {
+		char *xattr_name;
+		xattr_name = kmem_asprintf("%s%s", XATTR_USER_PREFIX, name);
```

Comment:
Since we're effectively redefining the on-disk format for user xattrs names with this change we can define what that should be.  Have you given any thought to making the prefixed "user." version the new format instead?  If all xattrs includes a namespace prefix the names would be consistent and less ambiguous.

Reply:
I have considered it, and while I do appreciate the approach for its cleanliness and internal consistency, I went in the direction of consistency with the existing implementations outside of Linux. There is a touch of aversion to the extra asprintfs required for the prefix as well, but I haven't actually attempted to measure any performance impact that might have.

Reply:
Well and the other argument for no-prefixes I can sympathize with is it does make the xattrs slightly smaller which means they're a little more likely to fit in the dnode SA space and not spill.  When that happens it does save a read I/O which is good for performance.